cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -0,0 +1,515 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.artifact;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.DatasetDefinition;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scan;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.api.templates.plugins.PluginClass;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.io.Locations;
+import co.cask.cdap.common.namespace.NamespacedLocationFactory;
+import co.cask.cdap.data2.datafabric.dataset.DatasetsUtil;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.tx.DatasetContext;
+import co.cask.cdap.data2.dataset2.tx.Transactional;
+import co.cask.cdap.proto.Id;
+import co.cask.tephra.TransactionExecutor;
+import co.cask.tephra.TransactionExecutorFactory;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Supplier;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.io.ByteStreams;
+import com.google.gson.Gson;
+import com.google.inject.Inject;
+import org.apache.twill.filesystem.Location;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class manages artifacts as well as metadata for each artifact. Artifacts and their metadata cannot be changed
+ * once they are written, with the exception of snapshot versions. An Artifact can contain
+ * plugin classes and/or application classes. We may want to extend this to include other types of classes, such
+ * as datasets.
+ *
+ * Every time an artifact is added, the artifact contents are stored at a location based on its id:
+ * /namespaces/{namespace-id}/artifacts/{artifact-name}/{artifact-version}
+ *
+ * Several writes are also performed on the meta table, one for metadata about the artifact itself, and one
+ * for each entity contained in the artifact.
+ *
+ * Artifact metadata is stored with r:{namespace}:{artifact-name} as the rowkey, {artifact-version} as the column,
+ * and ArtifactMeta as the value.
+ *
+ * PluginClass metadata is stored with p:{namespace}:{plugin-type}:{plugin-name} as the rowkey,
+ * {artifact-version}:{artifact-name} as the column, and PluginClass as the value
+ *
+ * TODO: (CDAP-2764) add this part when we have a better idea of what needs to be in AppClass.
+ * AppClass metadata is stored with a:{namespace}:{app-name} as the rowkey,
```

Comment:
yeah they are different because we need to be able to look up plugins on their own.  We may not need to be able to look up applications on their own just yet though, since currently an artifact contains a single application.